master/pkg/model/metrics.go

```diff
@@ -19,6 +19,11 @@ const (
 	InferenceMetricGroup MetricGroup = "inference"
 )
 
+// ProfilingMetricGroups designates metrics from profiling runs.
+var ProfilingMetricGroups = []MetricGroup{
+	"gpu", "cpu", "memory", "disk", "network",
```

Comment:
I thought we could have a single group called `system_metrics`, and have all the various metrics like `gpu_utilization` in it. then, when someone wants to add infiniband, we'd just add `ib_somethingsomething` metric rather than creating a new category. basically, I don't want the master to be a limiting factor when we add new metrics on the client.

Reply:
so i went back and forth on this. originally having a single `system_metrics` group was my approach too. i ended up with these separate groups because of how i wanted the client metrics reports to be structured. i landed on a client that reports these metrics on a per-group <> timestamp basis as a tradeoff between accuracy of the timestamp reported and increased query time as a result of the number of rows generated from a single trial/profiling run.

with a single `system_metrics` group:
- the actual metric names (i.e. `gpu_util`) would be stored in the metrics json blob in the table, which means the web UI would have to query every single row's metric jsonb for a given trial ID to render one graph for GPU util. this is inefficient, especially since certain metric "groups" (like GPU) are associated with certain labels (GPU UUID) to be rendered as dropdown menus in the UI. 
- if the client reports each individual metric as a single row in the db (assuming system metrics are collected every 1 second for a 30 minute trial with 2 agents) we'd have 9 total system metrics * 2 agents * 60 * 30 =~ 32000 metrics for a single run. for GPU metrics on agents that have multiple GPUs, this number goes even higher.
- if we store all metrics together in a single row (as a "snapshot"), this number goes down, but query time will still be high due to having to traverse a larger jsonb. also we lose on the accuracy of the timestamp reported, since the time taken to collect all these metrics will be reduced down to a single timestamp.

by separating system metrics into individual groups:
- client reports metrics to API every timestamp per group (so `gpu` group would contain "gpu_util" and "gpu_memory" for each GPU UUID). this is logically consistent with client reporting: we use a single dependency/API for querying GPU metrics and collect all relevant GPU metrics at the same time.
- structurally this grouping seems natural given that separate groups might have separate "labels". `gpu->GPU-UUID-1->gpu_util => 0.10`, `gpu->GPU-UUID-2->gpu_util => 0.20`, etc
- queries for rendering metrics can be limited in scope to their "group" so we can render individual plots more efficiently. also given that certain groups have certain labels, this makes it easier to get all the available labels (GPU UUID for example) for the groups plots.

> I don't want the master to be a limiting factor when we add new metrics on the client.

this is a valid concern but probably not as limiting as it seems. the existing groups (cpu, gpu, network, disk, memory) should cover most if not all system metrics we'd want to report. if we do need to add a new group for whatever reason, the metrics data may be different enough structurally that a new group in the master/db queryable by the web UI might be a helpful distinction to have.

Reply:
good point, I'm sold